### DISCUSSION TITLE
How to generate/play audio in a widget?

### DISCUSSION THREAD

**Discussion by szabolcsnagy** [johnlindquist/kit#855] [Q&A] [answered]
My ultimate goal is to have a widget to play some audio notification after a certain period of time. 
In the browser I can generate sounds using the `window.AudioContext` api, but if I try that in the widget script it crashes and there is nothing helpful in the logs.
What are my options to generate some "noise" in a widget?

**Comment by johnlindquist** [comment, marked as answer]
Send an event back to the script and play the sounds using the system/node. The Kit SDK has a built-in `playAudioFile` which uses the system `afplay`:

```js
global.playAudioFile = async (path, playInBackground = true) => {
    return global.exec(`afplay ${path} ${playInBackground ? "&>/dev/null &" : ""}`);
};
```

So using `playAudioFile` would look like this:


[Open sound-widget in Script Kit](https://scriptkit.com/api/new?name=sound-widget&url=https://gist.githubusercontent.com/johnlindquist/cbf78031e95cc27be58f52a7e2c471de/raw/223cd0f51a63d2be827f4c1d8ba35cd8f0f592ef/sound-widget.js")

```js
// Name: Sound Widget

import "@johnlindquist/kit"

let soundsPath = `/System/Library/Sounds`
let sounds = await readdir(soundsPath)

let buttons = sounds.map((sound) => { 
    return `<button class="p-2 rounded border" id="${sound}">${sound}</button>`
}).join("")

let w = await widget(`
<div class="flex flex-wrap">
    ${buttons}
</div>
`)

w.onClick(async ({ targetId }) => { 
    let fullPath = path.resolve(soundsPath, targetId)
    await playAudioFile(fullPath)
})


```

Let me know if I need to clarify anything. Always happy to make more demos.

**Reply by szabolcsnagy** [reply]
Awesome. I didn't know about the `afplay` nor the sound library that just exists on my mac. You solved my next problem as well. 😄 
Thank you very much.
